File: core/src/main/java/com/mcreater/amclcore/util/SwingUtil.java
package com.mcreater.amclcore.util;

import com.mcreater.amclcore.concurrent.task.AbstractAction;
import com.mcreater.amclcore.concurrent.task.AbstractTask;
import com.mcreater.amclcore.i18n.Text;
import com.mcreater.amclcore.util.platform.OperatingSystem;

import java.awt.*;
import java.awt.datatransfer.StringSelection;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;

import static com.mcreater.amclcore.i18n.I18NManager.translatable;
import static com.mcreater.amclcore.util.StringUtil.shortString;

public class SwingUtil {
    private static final String[] linuxBrowsers = {
            "xdg-open",
            "google-chrome",
            "firefox",
            "microsoft-edge",
            "opera",
            "konqueror",
            "mozilla"
    };
    public enum BrowserState {
        SUCCESS,
        NOT_FOUND,
        INTERNAL_EXCEPTION
    }

    public static class BrowserOpenTask extends AbstractTask<BrowserState> {
        private final String url;

        private BrowserOpenTask(String url) {
            this.url = url;
        }

        protected BrowserState call() {
            try {
                if (supportAction(Desktop.Action.BROWSE)) {
                    getDesktop().browse(URI.create(url));
                } else {
                    if (OperatingSystem.CURRENT_OS == OperatingSystem.WINDOWS) {
                        Runtime.getRuntime().exec(new String[]{"rundll32.exe", "url.dll,FileProtocolHandler", url});
                    } else if (OperatingSystem.CURRENT_OS == OperatingSystem.LINUX) {
                        for (String browser : linuxBrowsers) {
                            try (final InputStream is = Runtime.getRuntime().exec(new String[]{"which", browser}).getInputStream()) {
                                if (is.read() != -1) {
                                    Runtime.getRuntime().exec(new String[]{browser, url});
                                    return BrowserState.SUCCESS;
                                }
                            } catch (Throwable e) {
                                return BrowserState.INTERNAL_EXCEPTION;
                            }
                        }
                    } else if (OperatingSystem.CURRENT_OS == OperatingSystem.OSX) {
                        Runtime.getRuntime().exec(new String[]{"/usr/bin/open", url});
                    } else {
                        return BrowserState.INTERNAL_EXCEPTION;
                    }
                }
            } catch (IOException e) {
                return BrowserState.INTERNAL_EXCEPTION;
            }
            return BrowserState.NOT_FOUND;
        }

        protected Text getTaskName() {
            return translatable("core.swing.api.task.browser.name", shortString(url, 20));
        }
    }

    public static class CopyContentTask extends AbstractAction {
        private final String content;

        private CopyContentTask(String content) {
            this.content = content;
        }

        protected void execute() {
            Toolkit.getDefaultToolkit()
                    .getSystemClipboard()
                    .setContents(new StringSelection(content), (clipboard1, transferable) -> {
                    });
        }

        protected Text getTaskName() {
            return translatable("core.swing.api.task.clipboard.name", shortString(content, 20));
        }
    }
    public static BrowserOpenTask openBrowserAsync(String url) {
        return new BrowserOpenTask(url);
    }

    public static CopyContentTask copyContentAsync(String content) {
        return new CopyContentTask(content);
    }


    private static boolean supportAction(Desktop.Action action) {
        return getDesktop().isSupported(action);
    }

    private static Desktop getDesktop() {
        return Desktop.getDesktop();
    }
}

File: core/src/main/java/com/mcreater/amclcore/util/SwingUtil.kt
package com.mcreater.amclcore.util

import com.mcreater.amclcore.concurrent.task.AbstractAction
import com.mcreater.amclcore.concurrent.task.AbstractTask
import com.mcreater.amclcore.i18n.I18NManager.Companion.translatable
import com.mcreater.amclcore.i18n.Text
import com.mcreater.amclcore.util.platform.OperatingSystem
import java.awt.Desktop
import java.awt.Toolkit
import java.awt.datatransfer.Clipboard
import java.awt.datatransfer.StringSelection
import java.awt.datatransfer.Transferable
import java.io.IOException
import java.net.URI


class SwingUtil {
    companion object {
        @JvmStatic
        private val linuxBrowsers = arrayOf(
            "xdg-open",
            "google-chrome",
            "firefox",
            "microsoft-edge",
            "opera",
            "konqueror",
            "mozilla"
        )

        @JvmStatic
        fun openBrowserAsync(url: String): BrowserOpenTask {
            return BrowserOpenTask(url)
        }

        @JvmStatic
        fun copyContentAsync(content: String): CopyContentTask {
            return CopyContentTask(content)
        }

        @JvmStatic
        private fun supportAction(action: Desktop.Action): Boolean {
            return desktop.isSupported(action)
        }

        @JvmStatic
        private val desktop: Desktop
            get() = Desktop.getDesktop()
    }

    enum class BrowserState {
        SUCCESS,
        NOT_FOUND,
        INTERNAL_EXCEPTION
    }

    class BrowserOpenTask(private val url: String) :
        AbstractTask<BrowserState?>() {
        override fun call(): BrowserState {
            try {
                if (supportAction(Desktop.Action.BROWSE)) {
                    desktop.browse(URI.create(url))
                } else {
                    when (OperatingSystem.CURRENT_OS) {
                        OperatingSystem.WINDOWS -> {
                            Runtime.getRuntime().exec(
                                arrayOf(
                                    "rundll32.exe", "url.dll,FileProtocolHandler",
                                    url
                                )
                            )
                        }

                        OperatingSystem.LINUX -> {
                            for (browser in linuxBrowsers) {
                                try {
                                    Runtime.getRuntime().exec(arrayOf("which", browser)).inputStream.use { `is` ->
                                        if (`is`.read() != -1) {
                                            Runtime.getRuntime().exec(arrayOf(browser, url))
                                            return BrowserState.SUCCESS
                                        }
                                    }
                                } catch (e: Throwable) {
                                    return BrowserState.INTERNAL_EXCEPTION
                                }
                            }
                        }

                        OperatingSystem.OSX -> {
                            Runtime.getRuntime().exec(arrayOf("/usr/bin/open", url))
                        }

                        else -> {
                            return BrowserState.INTERNAL_EXCEPTION
                        }
                    }
                }
            } catch (e: IOException) {
                return BrowserState.INTERNAL_EXCEPTION
            }
            return BrowserState.NOT_FOUND
        }

        override fun getTaskName(): Text {
            return translatable("core.swing.api.task.browser.name", StringUtil.shortString(url, 20))
        }
    }

    class CopyContentTask(private val content: String) : AbstractAction() {
        override fun execute() {
            Toolkit.getDefaultToolkit()
                .systemClipboard
                .setContents(StringSelection(content)) { _: Clipboard?, _: Transferable? -> }
        }

        override fun getTaskName(): Text {
            return translatable(
                "core.swing.api.task.clipboard.name", StringUtil.shortString(
                    content, 20
                )
            )
        }
    }
}
